### PR TITLE
Keyboard shortcuts

### PR DESCRIPTION
**Pull request by neuromantic99**
Added keyboard shortcuts to gui

### CODE REVIEW

**Comment by ThomasAkam** [comment]
Thanks for the contribution, I think keyboard shortcuts will be useful but that this is something we should proceed with quite conservatively to ensure that all keyboard shortcuts we add are useful and well thought out.  I am a bit concerned in the current proposal that the same key can mean different things on different tabs, e.g. 'd' can mean both disconnect and delete experiment, which could get quite confusing.  I am inclinded to implement a minimal set of keyboard shortcuts initially, and then add more later if we think they are justified.   In practice which ones do you find most useful?  I am also a bit concerned that because many shortcuts are single key presses rather than key combinations  people may accidentally trigger things when they don't mean to (e.g start an experiment by accident by pressing space when they were trying to do something else), do you find this is an issue in practice?

Also, please could you seperate out the bugfix from the shortcuts into seperate pull requests - I think it is easy to work with pull requests if they are each for single issues.  You can do this by creating seperate branches for each pull request.

**Comment by ThomasAkam** [comment]
@alustig3 Would welcome your input on this if you have any thoughts about whether and where keyboard shortcuts would be useful.

**Comment by alustig3** [comment]
I agree that it may better to add shortcuts slowly/as needed. If they are added, I think it would be nice to have a dialog accessible through the help menu that lists the available shortcuts. Bonus if it also had a checkbox that allows you to turn the shortcuts on/off.

Personally, I like and would use the following of the proposed shortcuts:

### Experiment save (Ctrl+s)
I think it’s pretty natural for users to assume ctrl+s or cmd+s saves things. Right now however there is no indication that something was just saved. I would recommend adding some visual cue (a popup dialog? a change in save button color?) that shows the latest changes were saved.

### Start/stop task (Spacebar)
I like the space bar for starting/stopping the task in the run task tab. I think it is intuitive, and would save time because I often use the run_task tab to develop/debug tasks, which involves a lot of starting/stopping/resetting.

### Data/Task folder opening (Ctrl+D, Ctrl+T)
This can actually be added to the menu action itself in pyContrul_GUI.py. It will automatically label the menu item with the shortcut.
Screen Shot 2020-01-07 at 11 27 54 AM
Screen Shot 2020-01-07 at 11 27 23 AM

File: gui/configure_experiment_tab.py [hunk 1]
```diff
@@ -95,6 +95,15 @@ def __init__(self, parent=None):
         self.save_button.clicked.connect(self.save_experiment)
         self.run_button.clicked.connect(self.run_experiment)
 
+        # Keyboard shortcuts (select data dir shortcut carried over from run_task_tab)
+        shortcut_dict = {
+                        'n' : (lambda: self.new_experiment(dialog=True)),
+                        'd' : (lambda: self.delete_experiment()),
+                        's' : (lambda: self.save_experiment()),
```
[the remark this answers is not in the document]
Comment (ThomasAkam) [reply gui/configure_experiment_tab.py]: Thanks for the comments.  I agree that having a dialog accessable via the help menu which lists the keyboard shortcuts is desirable, and like the idea of a checkbox in this dialog which allows shortcuts to be enabled/disabled.

I agree that ctrl+s is almost universally used for saveing things and should be used for save experiment.

I think that in the run task tab both the spacebar to start/stop the task, and the 'u' to upload/reset the task will be useful when prototyping tasks, but I'm not completely convinced we need shortcuts for the actions that are done infrequently like connecting/disconnecting from the board.

**Comment by neuromantic99** [comment]
I agree that it's best to carefully add keyboard shortcuts as required and am happy for most of them to be removed, the 'every button has a keyboard shortcut' commit was intended as a jumping off point, I think Andy's suggestion works well to start with.

Most of the shortcuts in run_task_tab are mainly useful for debugging tasks / the gui itself as it allows you to make changes and quickly test them without clicking around (currently c to connect, u to upload then Space to run and Ctrl-w to close and start again). These are probably not worth implementing for users though if e.g. someone tries to Ctrl-w a chrome tab and accidentally ends their experiment, so maybe some shortcuts could be exclusive to dev branches?

**Comment by ThomasAkam** [comment]
@neuromantic99 Do you think you will have time to make the proposed changes soon?  I would like to release annother version of pyControl as there is a lot on new functionallity and bugfixes since the last one.

**Comment by neuromantic99** [comment]
Sure thing, just so i'm clear, the shortcuts you'd like implemented are:

Experiment save (Ctrl+s)
Start/stop task (Spacebar)
Data/Task folder opening (Ctrl+D, Ctrl+T)

what about:
Upload (u) and any others?

**Comment by ThomasAkam** [comment]
I think (u) : upload / reset  on the run task tab is worth having.  Are there any others you feel strongly about?

**Comment by neuromantic99** [comment]
I've updated shortcuts to reflect suggestions and added a 'Saved Experiment' Dialog

**Comment by ThomasAkam** [comment]
I have implemented the help dialog listing the keyboard shortcuts.  I have not implemented the option of turning on/off the shortcuts as there did not appear to be a simple way to do this.  @neuromantic99  and @alustig3 could you take a look and see if you are happy for this to be merged into the dev branch.

**Comment by alustig3** [comment]
For me on a mac, I thought the help dialog text was too small. Also, it was possible for the window to be resized, and the shortcuts would be hidden.

Screen Shot 2020-01-14 at 6 25 37 PM
Screen Shot 2020-01-14 at 6 25 48 PM

Here is what my changes look like on a mac:
Screen Shot 2020-01-14 at 6 37 32 PM

I tried pushing the code, but I don't think I have permission to push to James's branch. I've pasted the new code below: 

``` python
class Keyboard_shortcuts_dialog(QtGui.QDialog):
    '''Dialog for displaying summary variables from an experiment as a table.
    The table is copied to the clipboard as a string that can be pasted into a
    spreadsheet.'''
    def __init__(self, parent):
        super(QtGui.QDialog, self).__init__(parent)
        self.setWindowTitle('Keyboard shortcuts')

        self.Vlayout = QtGui.QVBoxLayout(self)

        self.global_header = QtGui.QLabel('<b><u>Global:</u></b>')
        self.task_shortcut = QtGui.QLabel('<kbd style="color:#c40017;">Ctrl + t</kbd> : Open tasks folder')
        self.data_shortcut = QtGui.QLabel('<kbd style="color:#c40017;">Ctrl + d</kbd> : Open data folder')
        
        self.run_task_header =      QtGui.QLabel('<br></br><b><u>Run task tab:</u></b>')
        self.select_task_shortcut = QtGui.QLabel('<kbd style="color:#c40017;">    t    </kbd> : Select task')
        self.upload_shortcut =      QtGui.QLabel('<kbd style="color:#c40017;">    u    </kbd> : Upload/reset task')
        self.start_shortcut =       QtGui.QLabel('<kbd style="color:#c40017;">spacebar </kbd> : Start/stop task')

        self.experiment_header = QtGui.QLabel('<br></br><b><u>Experiments tab:</u></b>')
        self.save_shortcut =     QtGui.QLabel('<kbd style="color:#c40017;">Ctrl + s</kbd> : Save experiment ')

        for lbl in [self.global_header,self.task_shortcut,self.data_shortcut, self.run_task_header, self.select_task_shortcut,self.upload_shortcut,self.start_shortcut,self.experiment_header,self.save_shortcut]:
            lbl.setFont(QtGui.QFont('Helvetica', 12))
            self.Vlayout.addWidget(lbl)

        self.setFixedSize(self.sizeHint())
```

**Comment by ThomasAkam** [comment]
Thanks, using labels rather than a textbox to ensure correct size on different systems is a good idea I have integrated in the last commit.  On windows the text at 12 point is quite a bit larger than other GUI text so I reduced it to 10, and the kbd font is very thin so I just used standard bold.  Could you check it looks OK on mac, then I will merge into the dev branch.

**Comment by alustig3** [comment]
The window title is cutoff for me, but I'm not sure how to prevent that other than hardcoding a larger setFixedSize(). A hacky workaround would be to not give a window title and then add a title label. Also, I liked having the shortcuts be a different color, not a big deal though.

before:
Screen Shot 2020-01-15 at 11 28 03 AM

after:
Screen Shot 2020-01-15 at 11 26 59 AM


``` python
class Keyboard_shortcuts_dialog(QtGui.QDialog):
    '''Dialog for displaying information about keyboard shortcuts.'''
    def __init__(self, parent):
        super(QtGui.QDialog, self).__init__(parent)

        self.Vlayout = QtGui.QVBoxLayout(self)

        label = QtGui.QLabel('<center><b>Keyboard Shortcuts</b></center<br></br>')
        label.setFont(QtGui.QFont('Helvetica', 12))
        self.Vlayout.addWidget(label)

        label_strings = [
            '<b><u>Global:</u></b>',
            '<b style="color:#0220e0;">Ctrl + t</b> : Open tasks folder',
            '<b style="color:#0220e0;">Ctrl + d</b> : Open data folder',
        
            '<br></br><b><u>Run task tab:</u></b>',
            '<b style="color:#0220e0;">    t    </b> : Select task',
            '<b style="color:#0220e0;">    u    </b> : Upload/reset task',
            '<b style="color:#0220e0;">spacebar </b> : Start/stop task',

            '<br></br><b><u>Experiments tab:</u></b>',
            '<b style="color:#0220e0;">Ctrl + s</b> : Save experiment ']

        for ls in label_strings:
            label = QtGui.QLabel(ls)
            label.setFont(QtGui.QFont('Helvetica', 10))
            self.Vlayout.addWidget(label)

        self.setFixedSize(self.sizeHint())
```

**Comment by ThomasAkam** [comment]
Have integrated Andy's latest code and merged the pull request into the dev branch.  Thanks both for the contribution.

**Pull request merged**